--- origsrc/avian/makefile	2009-04-11 12:55:35.000000000 -0500
+++ src/avian/makefile	2009-11-30 20:34:22.599777500 -0600
@@ -374,7 +384,7 @@ $(classpath-build)/%.class: $(classpath)
 $(classpath-dep): $(classpath-sources)
 	@echo "compiling classpath classes"
 	@mkdir -p $(dir $(@))
-	$(javac) -d $(dir $(@)) -bootclasspath $(classpath-build) \
+	$(javac) -source 1.5 -target 1.5 -d $(dir $(@)) -bootclasspath $(classpath-build) \
 		$(shell $(MAKE) -s --no-print-directory $(classpath-classes))
 	@touch $(@)
 
@@ -384,9 +394,9 @@ $(test-build)/%.class: $(test)/%.java
 $(test-dep): $(test-sources)
 	@echo "compiling test classes"
 	@mkdir -p $(dir $(@))
-	$(javac) -d $(dir $(@)) -bootclasspath $(classpath-build) \
+	$(javac) -source 1.5 -target 1.5 -d $(dir $(@)) -bootclasspath $(classpath-build) \
 		$(shell $(MAKE) -s --no-print-directory $(test-classes))
-	$(javac) -source 1.2 -target 1.1 -XDjsrlimit=0 -d $(dir $(@)) \
+	$(javac) -source 1.3 -target 1.2 -XDjsrlimit=0 -d $(dir $(@)) \
 		test/Subroutine.java
 	@touch $(@)
 
