--- origsrc/avian/classpath/jni-util.h	2009-04-11 11:32:27.000000000 -0500
+++ src/avian/classpath/jni-util.h	2009-11-30 20:28:48.885396500 -0600
@@ -15,7 +15,7 @@
 #include "stdlib.h"
 
 #undef JNIEXPORT
-#ifdef __MINGW32__
+#if defined(__MINGW32__) || defined(__CYGWIN__)
 #  define JNIEXPORT __declspec(dllexport)
 #else
 #  define JNIEXPORT __attribute__ ((visibility("default")))
--- origsrc/avian/makefile	2009-04-11 12:55:35.000000000 -0500
+++ src/avian/makefile	2009-11-30 20:34:22.599777500 -0600
@@ -14,7 +14,7 @@ build-platform = \
 		| sed 's/^cygwin.*$$/cygwin/')
 
 arch = $(build-arch)
-platform = $(subst cygwin,windows,$(subst mingw32,windows,$(build-platform)))
+platform = $(subst mingw32,windows,$(build-platform))
 
 ifeq ($(platform),windows)
 	arch = i386
@@ -126,6 +126,16 @@ ifeq ($(platform),darwin)
 	shared = -dynamiclib
 endif
 
+ifeq ($(platform),cygwin)
+	object-format = pe-i386
+
+	so-prefix = cyg
+	so-suffix = .dll
+	exe-suffix = .exe
+
+	cflags = $(common-cflags)
+endif
+
 ifeq ($(platform),windows)
 	inc = "$(root)/win32/include"
 	lib = "$(root)/win32/lib"
@@ -449,7 +459,7 @@ endif
 $(generator-objects): $(native-build)/%.o: $(src)/%.cpp
 	@echo "compiling $(@)"
 	@mkdir -p $(dir $(@))
-	$(build-cxx) -DPOINTER_SIZE=$(pointer-size) -O0 -g3 $(build-cflags) \
+	$(build-cxx) -DPOINTER_SIZE=$(pointer-size) -O0 -g3 $(common-cflags) \
 		-c $(<) -o $(@)
 
 $(jni-objects): $(native-build)/%.o: $(classpath)/%.cpp
--- origsrc/avian/src/boot.cpp	2009-04-11 11:32:27.000000000 -0500
+++ src/avian/src/boot.cpp	2009-11-30 20:28:48.885396500 -0600
@@ -17,7 +17,7 @@ extern "C" void __cxa_pure_virtual(void)
 
 #ifdef BOOT_IMAGE
 
-#ifdef __MINGW32__
+#if defined(__MINGW32__) || defined(__CYGWIN__)
 #  define EXPORT __declspec(dllexport)
 #  define SYMBOL(x) binary_bootimage_bin_##x
 #else
@@ -43,7 +43,7 @@ extern "C" {
 
 #ifdef BOOT_CLASSPATH
 
-#ifdef __MINGW32__
+#if defined(__MINGW32__) || defined(__CYGWIN__)
 #  define EXPORT __declspec(dllexport)
 #  define SYMBOL(x) binary_classpath_jar_##x
 #else
--- origsrc/avian/src/common.h	2009-04-11 11:32:27.000000000 -0500
+++ src/avian/src/common.h	2009-11-30 20:28:48.900996500 -0600
@@ -24,6 +24,9 @@
 #ifdef __MINGW32__
 #  define JNIEXPORT __declspec(dllexport)
 #  define PATH_SEPARATOR ';'
+#elif defined __CYGWIN__
+#  define JNIEXPORT __declspec(dllexport)
+#  define PATH_SEPARATOR ':'
 #else
 #  define JNIEXPORT __attribute__ ((visibility("default")))
 #  define PATH_SEPARATOR ':'
@@ -50,13 +53,15 @@
 
 #ifdef __MINGW32__
 #  define SO_PREFIX ""
+#elif defined(__CYGWIN__)
+#  define SO_PREFIX "cyg"
 #else
 #  define SO_PREFIX "lib"
 #endif
 
 #ifdef __APPLE__
 #  define SO_SUFFIX ".jnilib"
-#elif defined __MINGW32__
+#elif defined(__MINGW32__) || defined(__CYGWIN__)
 #  define SO_SUFFIX ".dll"
 #else
 #  define SO_SUFFIX ".so"
