--- origsrc/avian/src/posix.cpp	2009-12-05 16:41:08.000000000 -0600
+++ src/avian/src/posix.cpp	2010-10-26 23:26:44.063972100 -0500
@@ -25,7 +25,9 @@
 #include "unistd.h"
 #include "pthread.h"
 #include "signal.h"
+#ifndef __CYGWIN__
 #include "sys/ucontext.h"
+#endif
 #include "stdint.h"
 #include "dirent.h"
 
@@ -802,6 +804,7 @@ class MySystem: public System {
 void
 handleSignal(int signal, siginfo_t* info, void* context)
 {
+#ifndef __CYGWIN__
   ucontext_t* c = static_cast<ucontext_t*>(context);
 
 #ifndef BASE_REGISTER
@@ -812,10 +815,12 @@ handleSignal(int signal, siginfo_t* info
   void* base = reinterpret_cast<void*>(BASE_REGISTER(c));
   void* stack = reinterpret_cast<void*>(STACK_REGISTER(c));
   void* thread = reinterpret_cast<void*>(THREAD_REGISTER(c));
+#endif
 
   unsigned index;
 
   switch (signal) {
+#ifndef __CYGWIN__
   case VisitSignal: {
     index = VisitSignalIndex;
 
@@ -853,6 +858,7 @@ handleSignal(int signal, siginfo_t* info
       vmJump(ip, base, stack, thread, 0, 0);
     }
   } break;
+#endif
 
   case InterruptSignal: {
     index = InterruptSignalIndex;
